// @supercollider/dryads/lib/AudioBus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const dryadic_1 = require("dryadic");
/**
 * Allocates an audio bus, making it available in the children's context as .out (integer)
 * and .numChannels (integer)
 */
class AudioBus extends dryadic_1.Dryad {
    defaultProperties() {
        return {
            numChannels: 1,
        };
    }
    /**
     * If there is no SCServer in the parent context,
     * then this will wrap itself in an SCServer
     */
    requireParent() {
        return "SCServer";
    }
    prepareForAdd() {
        return {
            callOrder: "SELF_THEN_CHILDREN" /* SELF_THEN_CHILDREN */,
            updateContext: (context, properties) => ({
                out: context.scserver.state.allocAudioBus(properties.numChannels),
                numChannels: properties.numChannels,
            }),
        };
    }
    remove() {
        return {
            run: (context, properties) => context.scserver.state.freeAudioBus(context.out, properties.numChannels),
        };
    }
}
exports.default = AudioBus;
