// @supercollider/dryads/lib/SynthControl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const server_1 = require("@supercollider/server");
const dryadic_1 = require("dryadic");
const lodash_1 = tslib_1.__importDefault(require("lodash"));
const { nodeSet } = server_1.msg;
/**
 * Sends nodeSet messages to the Synth in the parent context.
 *
 * This takes a Bacon.js stream which should return objects
 * {param: value, ...} and sends `nodeSet` messages to the parent Synth.
 *
 * SynthControl should be a child of a Synth, Group or other Dryad that
 * sets context.nodeID
 */
class SynthControl extends dryadic_1.Dryad {
    /**
     * If there is no SCServer in the parent context,
     * then this will wrap itself in an SCServer
     */
    requireParent() {
        return "SCServer";
    }
    add(player) {
        return {
            run: (context, properties) => {
                if (properties.stream) {
                    const subscription = properties.stream.subscribe(event => {
                        // This assumes a Bacon event.
                        // Should validate that event.value is object
                        const msg = nodeSet(context.nodeID || -1, event.value());
                        player.callCommand(context.id, {
                            scserver: {
                                bundle: {
                                    time: 0.03,
                                    packets: [msg],
                                },
                            },
                        });
                    });
                    player.updateContext(context, { subscription });
                }
            },
        };
    }
    remove() {
        return {
            run: (context) => {
                if (context.subscription) {
                    if (lodash_1.default.isFunction(context.subscription)) {
                        // baconjs style
                        context.subscription();
                    }
                    else {
                        // Rx style
                        context.subscription.dispose();
                    }
                }
            },
        };
    }
}
exports.default = SynthControl;
