// @supercollider/dryads/lib/SynthEventList.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const server_1 = require("@supercollider/server");
const dryadic_1 = require("dryadic");
const lodash_1 = tslib_1.__importDefault(require("lodash"));
const Group_1 = tslib_1.__importDefault(require("./Group"));
const iterators_1 = require("./utils/iterators");
const { AddActions, synthNew } = server_1.msg;
/**
 * Takes a list of synth event objects with relative times and schedules them.
 *
 * #### properties
 *
 * __events:__ Array
 *
 * The event values should be simple JavaScript objects:
 *
 *     {
 *       defName: 'synthDefName',
 *       args: {
 *         out: 0,
 *         freq: 440
 *       },
 *       time: 0.3
 *     }
 *
 *  Where time is seconds relative to the epoch. The epoch is the start time of
 *  the dryadic tree, unless a parent Dryad has set a new epoch into context.
 *
 *    epoch: number|Date|undefined
 *      Optional epoch that the event times in the list are relative to.
 *      Can also be updated by the updateStream
 *      default: context.epoch or now
 *
 * __updateStream:__ Bacon stream to push updated event lists of the form:
 *
 *      {
 *        events: [{time: msgs: []}...],
 *        epoch: 123456789
 *      }

 *     .events Array
 *     .epoch  number|Date
 *
 * Deprecated: will be replaced with live updating and setting of
 * Any value in a dryadic document from the player or remote client.
 *
 * Pushing a new event list cancels previous events and schedules new events.
 *
 * Note that by default the epoch will be unchanged: relative times
 * are still relative to when the Dryad tree started playing or when any parent
 * Dryad set an epoch in context. This means you update the currently playing score
 * but it doesn't restart from the beginning, it keeps playing.
 *
 * Optionally you may push an .epoch with the updateStream. This can be a date or timestamp
 * slightly in the future. If you pass "now" then any events at `0.0` will be too late to play.
 *
 * __defaultParams:__ a fixed object into which the event value is merged.
 *
 * __loopTime:__ Play the events continuously in a loop.
 */
class SynthEventList extends dryadic_1.Dryad {
    defaultProperties() {
        return { events: [] };
    }
    add(player) {
        let commands = {
            scserver: {
                schedLoop: (context, properties) => {
                    // temporary: we need to know the play time of the whole document
                    const epoch = context.epoch || lodash_1.default.now() + 200;
                    if (epoch !== context.epoch) {
                        context = player.updateContext(context, { epoch });
                    }
                    // epoch, was 3rd arg
                    return this._makeSchedLoop(properties.events || [], properties.loopTime, context);
                },
            },
        };
        // built-in stream support will be added to Dryadic
        // for now it is hard to detect Bacon.Bus as being an object,
        if (this.properties.updateStream) {
            commands = lodash_1.default.assign(commands, {
                run: (context, properties) => {
                    const subscription = properties.updateStream.subscribe(streamEvent => {
                        const ee = streamEvent.value();
                        const loopTime = lodash_1.default.isUndefined(ee.loopTime) ? properties.loopTime : ee.loopTime;
                        const epoch = ee.epoch || context.epoch || lodash_1.default.now() + 200;
                        if (epoch !== context.epoch) {
                            context = player.updateContext(context, {
                                epoch,
                            });
                        }
                        player.callCommand(context.id, {
                            scserver: {
                                // need to set epoch as well because OSCSched uses that for relative times
                                schedLoop: (ctx /*, props*/) => this._makeSchedLoop(ee.events || [], loopTime, ctx),
                            },
                        });
                    });
                    player.updateContext(context, { subscription });
                },
            });
        }
        return commands;
    }
    _makeSchedLoop(events, loopTime, context) {
        const synthEvents = this._makeMsgs(events, context);
        return loopTime ? iterators_1.loopedEventListIterator(synthEvents, loopTime) : iterators_1.eventListIterator(synthEvents);
    }
    _makeMsgs(events, context) {
        const defaultParams = this.properties.defaultParams || {};
        return events.map(event => {
            // TODO: do this a jit time in the schedLoop
            const defName = event.defName || defaultParams.defName;
            const args = lodash_1.default.assign({ out: context.out || 0 }, defaultParams.args, event.args);
            const msg = synthNew(defName, -1, AddActions.TAIL, context.group, args);
            return {
                time: event.time,
                msgs: [msg],
            };
        });
    }
    remove() {
        return {
            run: (context) => {
                if (context.subscription) {
                    if (lodash_1.default.isFunction(context.subscription)) {
                        // baconjs style
                        context.subscription();
                    }
                    else {
                        // Rx style
                        context.subscription.dispose();
                    }
                }
            },
            scserver: {
                sched: (context) => {
                    // unschedAll
                    return this._makeSchedLoop([], undefined, context);
                },
            },
        };
    }
    /**
     * @return {Dryad}  Wraps itself in a Group so all child Synth events will be removed on removal of the Group.
     */
    subgraph() {
        return new Group_1.default({}, [this]);
    }
}
exports.default = SynthEventList;
